src/bridge/bridge.c

```diff
@@ -682,12 +685,33 @@ main (int argc,
   signal (SIGABRT, cockpit_test_signal_backtrace);
   signal (SIGSEGV, cockpit_test_signal_backtrace);
 
-  /* In case we are run via sshd and we have journald, make sure all
-   * logging output ends up in the journal on *this* machine, not sent
-   * back to the client.
-   */
-  if (g_getenv ("SSH_CONNECTION") && !g_log_writer_is_journald (2) && ! isatty(2))
+  if (g_strv_contains ((const gchar **) argv, "--privileged"))
+    {
+      /* We are being spawned, under sudo or pkexec, by the user's copy
+       * of the bridge.  In that case, the first thing that will happen,
+       * if we receive our stderr via the socket that is our stdin.
+       */
+      const char *msg = "\n{\"command\": \"send-stderr\"}";
+      g_print ("%zu\n%s", strlen (msg), msg);
+
+      g_autoptr(GUnixConnection) connection = cockpit_socket_connection_new_dup_fd (G_TYPE_UNIX_CONNECTION,
+                                                                                    STDIN_FILENO, &error);
+      if (!connection)
+        errx (EXIT_FAILURE, "couldn't create socket from stdin: %s", error->message);
+
+      int fd = g_unix_connection_receive_fd (connection, NULL, &error);
+      if (fd == -1)
+        errx (EXIT_FAILURE, "couldn't receive stderr socket from stdin: %s", error->message);
+
+      int r = dup2 (fd, 2);
```

Comment:
The commit message says "The first thing we do (before calling into any part of GLib) is to
request and receive the stderr file descriptor of the bridge which is
spawning us." but there is a lot of glib stuff going on here before calling `dup2`, no?

Reply:
That's a pretty good point...

Reply:
So basically, glib doesn't really consider that stderr might be replaced while the program is running.  It contains a static variable which is initialised at the first time that a message is printed (note: this does *not* exclude debug messages not shown) which records if stderr is the journal.  After that gets set, it's impossible to change it.

So yes, if we ran this with debugging output enabled, I guess we'd get in trouble.  But we'd also end up with a lot of debugging output in the sudo error message :)

Reply:
I addressed this by rewriting in pure C.